pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/FieldIndexConfigs.java

```diff
@@ -0,0 +1,131 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pinot.segment.spi.index;
+
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.apache.pinot.spi.config.table.IndexConfig;
+import org.apache.pinot.spi.utils.JsonUtils;
+
+
+/**
+ * FieldIndexConfigs are a map like structure that relates index types with their configuration, providing a type safe
+ * interface.
+ *
+ * This class can be serialized into a JSON object whose keys are the index type ids using Jackson, but cannot be
+ * serialized back. A custom Jackson deserializer could be provided if needed.
+ */
+public class FieldIndexConfigs {
+
+  public static final FieldIndexConfigs EMPTY = new FieldIndexConfigs(new HashMap<>());
+
+  private final Map<IndexType, IndexConfig> _configMap;
+
+  private FieldIndexConfigs(Map<IndexType, IndexConfig> configMap) {
+    _configMap = Collections.unmodifiableMap(configMap);
+  }
+
+  /**
+   * Returns the configuration associated with the given index type, which will be null if there is no configuration for
+   * that index type.
+   */
+  @JsonIgnore
+  public <C extends IndexConfig, I extends IndexType<C, ?, ?>> C getConfig(I indexType) {
+    IndexConfig config = _configMap.get(indexType);
+    if (config == null) {
+      return indexType.getDefaultConfig();
+    }
+    return (C) config;
+  }
+
+  /*
+  This is used by Jackson when this object is serialized. Each entry of the map will be directly contained in the
+  JSON object, with the key name as the key in the JSON object and the result of serializing the key value as the value
+  in the JSON object.
+   */
+  @JsonAnyGetter
+  public Map<String, JsonNode> unwrapIndexes() {
+    Function<Map.Entry<IndexType, IndexConfig>, JsonNode> serializer =
+        entry -> entry.getKey().serialize(entry.getValue());
+    return _configMap.entrySet().stream()
+        .filter(e -> e.getValue() != null)
+        .collect(Collectors.toMap(entry -> entry.getKey().getId(), serializer));
+  }
+
+  @Override
+  public String toString() {
+    try {
+      return JsonUtils.objectToString(this);
```

Comment:
We may remove the json annotations with the following change:
```suggestion
      return JsonUtils.objectToString(unwrapIndexes());
```

Reply:
I really prefer to add the jackson annotation, but given that it seems that you really prefer to do not add it, I will remove it 😆